nova/virt/vmwareapi/vmops.py

```diff
@@ -2058,13 +2131,69 @@ def migrate_disk_and_power_off(self, context, instance, dest, flavor):
                                        total_steps=RESIZE_TOTAL_STEPS)
 
         # 1. Power off the instance
-        vm_util.power_off_instance(self._session, instance, vm_ref)
         self._update_instance_progress(context, instance,
                                        step=1,
                                        total_steps=RESIZE_TOTAL_STEPS)
+        vm_util.power_off_instance(self._session, instance, vm_ref)
 
-    def confirm_migration(self, migration, instance, network_info):
+        # 2. Detach the volumes
+        self._update_instance_progress(context, instance,
+                                       step=2,
+                                       total_steps=RESIZE_TOTAL_STEPS)
+        self._detach_volumes(instance, block_device_info)
+
+        # 3. Detach network interfaces
+        self._update_instance_progress(context, instance,
+                                       step=3,
+                                       total_steps=RESIZE_TOTAL_STEPS)
+        device_change = self._get_network_device_change(
+            vm_ref, instance.image_meta, network_info, operation="remove")
+        vm_util.vm_device_change(self._session, vm_ref, device_change)
+
+        # 4. Copy the VM to the dest
+        self._update_instance_progress(context, instance,
+                                       step=4,
+                                       total_steps=RESIZE_TOTAL_STEPS)
+        self._do_migrate_disk(context, vm_ref, instance, dest_data)
```

Comment:
In general, shouldn't we have a revert for every step we did here? let's say the network detach failed, then cinder still sees the volumes as attached and the instance is in error. Customer will reset state and maybe try to power on again. We then have an incidient, because nobody get's the volumes back to the instance.

If we would try to attach them again, the instance would be usable again. Same for network, if the migrate fails.

Reply:
I've added a catch around this function, and we revert the networking, volumes and power state if anything fails in `_do_migrate_disk`.